modyn/selector/internal/selector_strategies/downsampling_strategies/scheduler_downsampling_strategy.py

```diff
@@ -0,0 +1,97 @@
+from modyn.selector.internal.selector_strategies.downsampling_strategies import AbstractDownsamplingStrategy
+from modyn.selector.internal.selector_strategies.downsampling_strategies.utils import instantiate_downsampler
+
+
+class SchedulerDownsamplingStrategy(AbstractDownsamplingStrategy):
+    def __init__(self, downsampling_config: dict, maximum_keys_in_memory: int):
+        if "first_strategy" not in downsampling_config:
+            raise ValueError(
+                "If you want to use a Downsampling Scheduler, please specify the first and the second strategy."
+            )
+        first_strategy = downsampling_config["first_strategy"]
+
+        if "second_strategy" not in downsampling_config:
+            raise ValueError(
+                "If you want to use a Downsampling Scheduler, please specify the first and the second strategy."
+            )
+        second_strategy = downsampling_config["second_strategy"]
+
+        if "threshold" not in downsampling_config:
+            raise ValueError("Please specify after how many epochs you want to switch strategy.")
+        self.threshold = downsampling_config["threshold"]
+
+        if "ratio" in downsampling_config:
+            raise ValueError(
+                "SchedulerDownsamplingStrategy has no downsampling ratio. Specify the downsampling ratio"
+                "inside first_strategy and second_strategy."
+            )
+        downsampling_config["ratio"] = 100  # just to deal with super
+
+        if "sample_then_batch" in downsampling_config:
+            raise ValueError("SchedulerDownsamplingStrategy has no downsampling mode (sample_then_batch parameter).")
+        downsampling_config["sample_then_batch"] = True  # useless, just to avoid ValueErrors
```

Comment:
mhm, I am not a big fan of this, it shows that somewhere the design could be improved, because it shows the inheritance hierarchy is not optimal, but I thought about it a bit and did not think of a better solution just yet

Reply:
I too have thought about it for a while but this seems the best solution. The problem is that this class (and also Empty) are not real downsamplers. The correct thing would probably be to have a `TrueAbstractDownsamplingStrategy(AbstractDownsamplingStrategy)` class which is responsible of sample_then_batch, ratios and so forth. And then leave the AbstractDownsamplingStrategy with almost no methods. But idk if it's worth, what do you think?

Reply:
Could you please add something like this:

```python
# In the current design, the SchedulingStrategy implements a DownsamplingStrategy, which by definition expects a concrete strategy as its subclasses, i.e., each downsampling strategy should have certain properties, such as sample_then_batch. The SchedulingStrategy does not fit this design, since itself it is not a strategy, but an object choosing a strategy. If this causes further issues, the canonical refactoring would be that _every_ downsampling strategy uses a scheduler, but some just never change the behavior
```

Reply:
so basically a comment explaining the problem and the solution, which, in my opinion, would be to have everything as a schedulingstrategy with potentially just one downsampler